File: src/348. Sum of a square and a cube/348.rb
class Integer
  def is_palindrome?
    self.to_s == self.to_s.reverse
  end
end

# if the result is not enough, take a bigger guess
upper_limit = 10 ** 9

squares = []
2.upto(Math.sqrt(upper_limit).floor) do |a|
  squares << a * a
end

cubes = []
b = 2
while b * b * b < upper_limit
  cubes << b * b * b
  b += 1
end

candidates = Hash.new(0)
squares.each do |s|
  cubes.each do |c|
    candidates[s + c] += 1 if (s + c).is_palindrome?
  end
end

candidates.keep_if {|k, v| v == 4}
if candidates.size >= 5
  puts candidates.keys.min(5).inject(:+)
end

File: src/348. Sum of a square and a cube/348.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <map>
#include <numeric>
#include <set>
#include <string>
#include <vector>

using std::map;
using std::set;
using std::string;
using std::vector;

static bool is_palindrome(long long n);

int main() {
    // if the result is not enough, take a bigger guess
    long long upper_limit = 1000000000LL;

    set<long long> squares;
    for (long long a = 2; a < std::sqrt(upper_limit); ++a) {
        squares.insert(a * a);
    }

    set<long long> cubes;
    for (long long b = 2; b * b * b < upper_limit; ++b) {
        cubes.insert(b *  b * b);
    }

    map<long long, int> candidates;
    for (long long s : squares) {
        for (long long c : cubes) {
            if (is_palindrome(s + c)) {
                ++candidates[s + c];
            }
        }
    }

    vector<long long> result;
    for (auto kv : candidates) {
        if (kv.second == 4) {
            result.push_back(kv.first);
        }
    }

    if (result.size() >= 5) {
        std::sort(result.begin(), result.end());
        std::cout << std::accumulate(result.begin(), result.begin() + 5, 0);
    }
}

static bool is_palindrome(long long n) {
    string original = std::to_string(n);
    string reversed(original.rbegin(), original.rend());
    return original == reversed;
}
